Fixed duration not display on youtube player.

waheedahmed opened the pull request:
[TNL-3056] (https://openedx.atlassian.net/browse/TNL-3056)

Fixed duration not displaying on youtube player, by adding ```moment``` as dependancy and fixed jasmine tests which were failing when we add ```moment``` as dependancy in video modules.

waheedahmed commented:
@singingwolfboy @muzaffaryousaf @cahrens @adampalay Please review.

cahrens commented:
:+1:

singingwolfboy commented on common/lib/xmodule/xmodule/js/main_requirejs.js:
```diff
@@ -0,0 +1,8 @@
+(function(requirejs) {
+    requirejs.config({
+        paths: {
+            "moment": "xmodule/include/common_static/js/vendor/moment.min"
```
It seems odd to have this file with just one entry in the path mapping. Is this really the only thing that needs to be configured here? Is there a different configuration file where this could be added, instead of creating this new file with one entry?

waheedahmed replied:
@singingwolfboy This is actually a mystery, I tried to find out a file like this which actually loading all other things in xmodule jasmine tests but couldn't find anything.

singingwolfboy replied:
@cahrens, do you know what's going on here? Is there a different file that @waheedahmed should be editing, instead?

cahrens replied:
The JS in common just depends on the application using it to load all of its dependencies. So for LMS, the dependencies are in /edx-platform/lms/static/lms/js/require-config.js, and for cms, /edx-platform/cms/static/cms/js/require-config.js.

Both of those already list moment. @waheedahmed does it work if you just declare the video dependency on moment in 01_initialize.js and leave out main_requirejs.js?

cahrens replied:
Oh, I see that main_requirejs.js is really only for tests. There should be someplace else to put that.

cahrens replied:
I don't think there is any requirejs config for tests in common. Like I mentioned, the common JS depends on the parent location, so the video code doesn't declare dependencies on third-party libraries like jquery, underscore, etc.

I think that what is going on is that moment realizes that it is being loaded in a RequireJS environment, and therefore doesn't put moment on the window object. This may be due to changes we made as part of the LMS optimizer workflow, that now moment is behaving differently, even for our namespaced version of RequireJS.

An option that may work is to use the "init" option in the shim configuration of RequireJS (in LMS and CMS)-- inside init, put moment on the window object. Then you could keep the code in 01_initialize.js as-is, and the tests would also work as-is.

waheedahmed replied:
@cahrens 399 Jasmine tests are failing for xmodule when we remove this ```main_requirejs.js``` and use ```moment``` as dependancy in ```01_initialize.js```.

waheedahmed replied:
@cahrens I tried ```init``` option in shim config but its not working for LMS and CMS.

muzaffaryousaf commented:
>I don't think there is any requirejs config for tests in common. Like I mentioned, the common JS depends on the parent location, so the video code doesn't declare dependencies on third-party libraries like jquery, underscore, etc.

@cahrens If i'm not wrong the requirejs config for tests in common lives [here](https://github.com/edx/edx-platform/blob/master/common/static/js/spec/main_requirejs.js) but as per my understanding they exist for tests that are in common but not in lib/xmodule. 

@waheedahmed Can you give it a try to add this new dependency in common requirejs config.

muzaffaryousaf commented:
@cahrens The interesting thing about the existing code `window.moment` is that sometime it gets load and undefined.

muzaffaryousaf commented on common/lib/xmodule/xmodule/js/src/video/01_initialize.js:
```diff
@@ -14,10 +14,9 @@
-
+['video/03_video_player.js', 'video/00_i18n.js', 'moment'],
+function (VideoPlayer, i18n, momentjs) {
+    var moment = momentjs || window.moment;
```
@waheedahmed I'm still not sure about this as it was giving error randomly on your devstack. Also i could't understand why the `window.moment`  gets undefined somewhere.  Did you know why the `window.moment ` gets undefined in existing code.

muzaffaryousaf replied:
@waheedahmed What if `momentjs` is not defined, would it load the video. Because you are expecting to load the moment from window object afterwards.

waheedahmed replied:
@muzaffaryousaf moment is used as dependancy here, so if it is undefined then video will not load.

muzaffaryousaf replied:
@waheedahmed Then `window.moment` should not be there.

cahrens commented on common/lib/xmodule/xmodule/js/main_requirejs.js:
```diff
@@ -0,0 +1,8 @@
+(function(requirejs) {
```
If we do end up introducing this file, please put it in the spec folder so it is clear that it is only for tests.

muzaffaryousaf commented on common/lib/xmodule/xmodule/js/js_test.yml:
```diff
@@ -40,6 +41,7 @@ lib_paths:
     - common_static/js/vendor/jasmine-imagediff.js
     - common_static/js/vendor/require.js
     - RequireJS-namespace-undefine.js
+    - main_requirejs.js
```
@cahrens FYI, we are adding this file in `lib_paths` so that it loads the moment before the actual video files.

waheedahmed commented:
>@cahrens If i'm not wrong the requirejs config for tests in common lives [here](https://github.com/edx/edx-platform/blob/master/common/static/js/spec/main_requirejs.js) but as per my understanding they exist for tests that are in common but not in lib/xmodule.

>@waheedahmed Can you give it a try to add this new dependency in common requirejs config.

@muzaffaryousaf this [main_requirejs.js] (https://github.com/edx/edx-platform/blob/master/common/static/js/spec/main_requirejs.js) is for tests, @cahrens was talking about ```require-config.js``` for LMS and CMS.
>An option that may work is to use the "init" option in the shim configuration of RequireJS (in LMS and CMS)-- inside init, put moment on the window object. Then you could keep the code in 01_initialize.js as-is, and the tests would also work as-is.

waheedahmed commented:
@andy-armstrong Please review this.

@cahrens @muzaffaryousaf Please take a second look.

waheedahmed commented:
jenkins run all

andy-armstrong commented on common/lib/xmodule/xmodule/js/src/video/01_initialize.js:
```diff
@@ -14,10 +14,9 @@
-['video/03_video_player.js', 'video/00_i18n.js'],
+['video/03_video_player.js', 'video/00_i18n.js', 'moment'],
 function (VideoPlayer, i18n) {
     var moment = window.moment;
```
Can you add ``moment`` as a third parameter and use it directly? Does it get shimmed so that it returns itself? That would be cleaner than using it via ``window``.

waheedahmed replied:
@andy-armstrong When we are adding it as third parameter its returning undefined, thats why I used it via window. I read somewhere i don't remember where that moment initialize itself on window object if we don't use ```noGlobal: true``` otherwise it will return undefined.

muzaffaryousaf replied:
@waheedahmed @andy-armstrong This is how moment works with requirejs http://momentjs.com/docs/#/use-it/require-js/

muzaffaryousaf replied:
>Note: To allow moment.js plugins to be loaded in requirejs environments, moment is created as a named module. Because of this, moment must be loaded exactly as as "moment", using paths to determine the directory. Requiring moment with a path like "vendor\moment" will return undefined.

@andy-armstrong The issue with our approach is that the global moment gets undefined sometime and to fix that we have to specify the moment in define that returns `undefined` but it sets the global moment on window object. Right @waheedahmed ?

waheedahmed replied:
@muzaffaryousaf Yes.

andy-armstrong replied:
Thanks for the explanation. I think we just need to add a shim like we do for all the other vendor libraries (see https://github.com/edx/edx-platform/blob/master/lms/static/lms/js/require-config.js#L90):

            "moment": {
                exports: "moment"
            },

That should cause it ``window.moment`` to be returned as the value of the third parameter, and then code doesn't need to access it through ``window``. 

If that doesn't work, then just leave this as is.

muzaffaryousaf replied:
@waheedahmed Can you please give it a try ?

cahrens replied:
I'm also confused as to whether or not we should have the following someplace, per http://momentjs.com/docs/#/use-it/require-js/

define(["moment"], function (moment) {
    moment().format();
});

andy-armstrong commented:
:+1: Looks good to me. I have a minor nit about using the dependency directly, but thumbs up either way.

andy-armstrong commented:
Christina and I discussed this and she has a better proposal. So don't merge this as it stands.

cahrens commented:
@andy-armstrong and I are also wondering how the moment JS will end up being downloaded, given the fact that the moment.min library is never loaded on the page. We are worried that it may end up being downloaded as a non-cacheable file (because it won't get the MD5 hash).

Can you try creating a sandbox with your solution and look in the JS debugger to see if moment is downloading as an individual file with no MD5 hash before the extension? If this is not the case, we need to understand how moment is getting downloaded.

However, if it does end up that moment is downloaded as a non-cacheable file, we think that it is necessary to add moment to the "application" part of the JS pipeline in lms/envs/common.py. LMS uses a mix of RequireJS and non-RequireJS code, so it is much more complicated than CMS; CMS should be fine with just the moment requirement being loaded via RequireJS.

waheedahmed commented:
@cahrens @andy-armstrong The moment is downloading without MD5 hash, so I added it into ```application``` part or ```PIPELINE_JS``` [here] (https://github.com/edx/edx-platform/blob/master/lms/envs/common.py#L1520) and added shim as @andy-armstrong suggested and now its working fine.
But there are three jasmine tests failing because moment is undefined in tests even after adding shim like this:

            "moment": {
                exports: "moment"
            }, 
in ```main_requirejs.js```.

cahrens commented:
Try adding the shim to spec/main_requirejs.js as well.

waheedahmed commented:
@cahrens I already tried it locally but it didn't work, lets see what happens on jenkins.

adampalay commented:
@waheedahmed , what's up with this PR?

waheedahmed commented:
@adampalay ```moment``` is not working correct in jasmine tests.

waheedahmed commented:
@cahrens We can fix this issue by adding following in 01_initialize.js:

```var moment = moment || window.moment;```

What do you think?

andy-armstrong commented:
@waheedahmed I think you just need to shim ``moment`` in the Jasmine RequireJS config as well:

https://github.com/edx/edx-platform/blob/master/lms/static/js/spec/main.js#L100

waheedahmed commented:
@andy-armstrong I don't think it will work because this ```main.js``` not loading for xmodule jasmine tests.

waheedahmed commented:
@andy-armstrong I tried it not working.

andy-armstrong commented:
@waheedahmed Sorry, my mistake. I didn't pick up that these were for xmodule. You should just need the shim in your new ``main_requirejs`` file.

andy-armstrong commented:
@waheedahmed It is a pain that we duplicate these RequireJS configuration files everywhere but I don't know of a better solution. You might also want to add the moment shim into the common ``main_requirejs.js`` file too, just to avoid confusion in the future.

waheedahmed commented:
@andy-armstrong I already added shim for moment in new ```main_requirejs``` and it's not working even after adding in common ```main_requirejs```.

waheedahmed commented:
@andy-armstrong Are we good now or you have any other idea?

andy-armstrong commented:
@waheedahmed I think the problem is that you haven't configured the xmodule tests to use RequireJS in the same way as other libraries do. For example, in common there is a separate RequireJS yml file:

https://github.com/edx/edx-platform/blob/master/common/static/js_test_requirejs.yml

It would be possible to do the same thing for xmodule but it probably isn't worth it. :+1: for this simple solution.

muzaffaryousaf commented:
:+1:
